scipy/spatial/distance.py

```diff
@@ -1497,10 +1497,10 @@ def squareform(X, force="no", checks=True):
         d = s[0]
 
         if d <= 1:
-            return np.array([], dtype=np.double)
+            return np.array([], dtype=X.dtype)
 
         # Create a vector.
-        v = np.zeros((d * (d - 1)) // 2, dtype=np.double)
+        v = np.zeros((d * (d - 1)) // 2, dtype=X.dtype)
 
         # Since the C code does not support striding using strides.
         # The dimensions are used instead.
```

Comment:
Good catch, fixed. I also updated the tests to cover the float32 case (and others).